### PR TITLE
VBM-61: Lot card component

### PR DESCRIPTION
**Pull request by mazurGit**
https://user-images.githubusercontent.com/69375560/187371276-ba587d45-a781-4621-adad-1dcf86f89482.mp4

### CODE REVIEW

File: packages/mobile/src/localization/languages/en.ts [hunk 1]
```diff
@@ -11,4 +11,19 @@ export const en = {
...
     LAST_NAME_HINT: 'Enter your last name',
   },
+  currency: {
+    UAH: 'UAH',
+  },
+  button: {
+    BID: 'Bid',
+    BUY: 'Buy',
+  },
+  time: {
```
Comment (seal323) [inline packages/mobile/src/localization/languages/en.ts:21]: I suppose we don't need it, cause we will use the library for if (F.e. dayjs) which supports localization. @AleksandrTermenzhy what do you think?

File: packages/mobile/src/screens/home/components/lot/lot.tsx [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    <View style={styles.container}>
+      <View style={styles.imgWrapper}>
+        <Image source={imgSrc} style={styles.img} />
+
+        {type === LotType.UPCOMING ? (
+          <View style={styles.time}>
+            <ClockIcon />
+            <Text style={globalStyles.ml2}>2 дня 5 г 15 хв</Text>
+          </View>
+        ) : null}
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/home/components/lot/lot.tsx:31]:
```suggestion
        {type === LotType.UPCOMING && (
          <View style={styles.time}>
            <ClockIcon />
            <Text style={globalStyles.ml2}>2 дня 5 г 15 хв</Text>
          </View>
        )}
```

File: packages/mobile/src/screens/home/components/lot/styles.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  divider: {
+    width: '100%',
+    height: 2,
+    backgroundColor: ColorPalette.GRAY_200,
+    marginVertical: 13,
+  },
+  priceWrapper: {
+    flexDirection: 'row',
+    alignItems: 'center',
+    justifyContent: 'space-between',
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/home/components/lot/styles.ts:66]: You can use globalStyles for these

**Review by seal323 — APPROVED** [review]

File: packages/mobile/src/screens/home/components/product/product.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  imageLinks,
+  title,
+  description,
+  price,
+  type,
+}) => {
+  const { t } = useTranslation();
+  const transformPrice = (price: string) => {
+    return price.length > 2 ? `${price.slice(0, 2)} ${price.slice(2)}` : price;
+  };
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/home/components/product/product.tsx:27]: Helper functions belong to `helpers` folder

File: packages/mobile/src/screens/home/components/product/product.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+  return (
+    <View style={[styles.container, globalStyles.boxShadow]}>
+      <View style={[styles.imgWrapper]}>
+        <Image source={{ uri: imageLinks[0] }} style={styles.img} />
+
+        {type === ProductType.AUCTION ? (
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/home/components/product/product.tsx:34]:
```suggestion
        {type === ProductType.AUCTION && (
```

File: packages/mobile/src/screens/home/components/product/product.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+          {title}
+        </Text>
+        <Text
+          style={[globalStyles.fs14, globalStyles.fontWeightRegular]}
+          numberOfLines={2}
+        >
+          {description}
+        </Text>
+        {type === ProductType.FIXED_PRICE && (
+          <Text style={globalStyles.fs12}>2 дні тому</Text>
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/home/components/product/product.tsx:63]: `2 дні тому` -- ?

File: packages/mobile/src/screens/home/components/product/product.tsx [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+          globalStyles.alignItemsCenter,
+          globalStyles.justifyContentSpaceBetween,
+        ]}
+      >
+        <Text
+          style={[globalStyles.fs16, globalStyles.fontWeightBold, styles.price]}
+        >{`${transformPrice(price)} ${t('common:currency.UAH')}`}</Text>
+        <Button
+          label={
+            type === ProductType.AUCTION
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/home/components/product/product.tsx:79]: I'd suggest extracting it to a variable: `type === ProductType.AUCTION`

File: packages/mobile/src/services/product/product.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+import { ApiRoutes, ProductDto } from '@vse-bude/shared';
+
+import { Http } from '~/services/http/http.service';
+
+type Constructor = {
+  http: Http;
+  apiPrefix: string;
+};
+
+class Product {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/product/product.ts:10]:
```suggestion
class ProductService {
```

File: packages/mobile/src/services/product/product.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  #apiPrefix: string;
+
+  constructor({ http, apiPrefix }: Constructor) {
+    this.#http = http;
+    this.#apiPrefix = apiPrefix;
+  }
+
+  getAllProducts(): Promise<ProductDto[]> {
+    return this.#http.load(`${this.#apiPrefix}${ApiRoutes.PRODUCTS}`, {
+      hasAuth: false,
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/product/product.ts:22]: Don't we need authentication for products API?

File: packages/mobile/src/store/products/actions.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+import { ActionType } from './common';
+
+const loadAllProducts = createAsyncThunk<
+  ProductDto[],
+  undefined,
+  AsyncThunkConfig
+>(ActionType.PRODUCTS_FETCH, async (_, { extra }) => {
+  const { productApi } = extra;
+
+  return await productApi.getAllProducts();
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/store/products/actions.ts:13]: You may skip `await` when returning from `async` function as it still returns a `Promise`
```suggestion
  return productApi.getAllProducts();
```

File: packages/mobile/src/store/products/reducer.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
+import { createReducer } from '@reduxjs/toolkit';
+import { DataStatus } from '~/common/enums/enums';
+import { ProductDto } from '@vse-bude/shared';
+import { loadAllProducts } from './actions';
+
+type InitialState = {
+  products: ProductDto[] | [];
+  productsLoadStatus: DataStatus;
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/store/products/reducer.ts:8]:
```suggestion
  dataStatus: DataStatus;
```

File: packages/shared/src/common/enums/product/index.ts [hunk 12]
```diff
@@ -0,0 +1,6 @@
+enum ProductType {
```
Comment (AleksandrTermenzhy) [inline packages/shared/src/common/enums/product/index.ts:1]: The file should be named `product.ts`

File: packages/shared/src/common/types/product/index.ts [hunk 13]
```diff
@@ -0,0 +1 @@
+export type { ProductDto } from './product-dto';
```
Comment (AleksandrTermenzhy) [inline packages/shared/src/common/types/product/index.ts:1]: Again, we don't use index-files. Pls, use folder name instead

File: packages/shared/src/common/types/product/product-dto.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+  minimalBid: string;
+  imageLinks: string[];
+  city: string;
+  type: ProductType;
+  status: string;
+  endDate: string;
+  cancellReason: string;
+  authorId: string;
+  categoryId: string;
+  winnerId: string | boolean;
```
Comment (AleksandrTermenzhy) [inline packages/shared/src/common/types/product/product-dto.ts:18]: Are you sure about that type?

**Review by AleksandrTermenzhy — CHANGES_REQUESTED** [review]

File: packages/mobile/src/store/products/selectors.ts [hunk 15]
```diff
@@ -0,0 +1,12 @@
+import { RootState } from '~/common/types/types';
+import { createSelector } from '@reduxjs/toolkit';
+
+const selectProductById = (id: string) =>
+  createSelector(
+    (state: RootState) => state.products.products,
+    (products) => {
+      return products.find((item) => item.id === id);
+    },
+  );
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/store/products/selectors.ts:10]: Please refer to this link and rewrite accordingly: https://github.com/reduxjs/reselect#q-how-do-i-create-a-selector-that-takes-an-argument

File: packages/mobile/src/screens/home/components/product/product.tsx [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+              globalStyles.flexDirectionRow,
+              globalStyles.alignItemsCenter,
+              globalStyles.justifyContentCenter,
+              { backgroundColor: colors.whiteColor, borderColor: colors.line },
+            ]}
+          >
+            <ClockIcon style={globalStyles.mr2} />
+            <Text style={globalStyles.fs12}>{duration}</Text>
+          </View>
+        )}
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/home/components/product/product.tsx:43]: Extract this to a separate component, pls

File: packages/mobile/src/helpers/date/get-time-to-event/get-time-to-event.helper.ts [hunk 17]
```diff
@@ -0,0 +1,26 @@
+import dayjs from 'dayjs';
+import { t } from 'i18next';
+
+const getTimeToEvent = (date: Date) => {
+  const duration = dayjs.duration(dayjs(date).diff(dayjs()));
+  if (duration.asMilliseconds() > 60000) {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/helpers/date/get-time-to-event/get-time-to-event.helper.ts:6]: No magic numbers please

File: packages/mobile/src/index.tsx [hunk 18]
```diff
@@ -8,10 +8,12 @@ import { store } from '~/store/store';
 import { NavigationDarkTheme, NavigationTheme } from '~/config/config';
 import { useColorScheme } from '~/hooks/hooks';
 import { i18 } from './localization/localization';
+import { dayjsInit } from './day-js/day-js-init';
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/index.tsx:11]: We don't need that root folder. Use helpers or create some sort of AppService with init() method instead

File: packages/mobile/src/helpers/helpers.ts [hunk 19]
```diff
@@ -1 +1,3 @@
 export { getQueryString } from './http/http';
+export { formatPrice } from './name-modifier/name-modifier';
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/helpers/helpers.ts:2]: What "name" are we talking about? Please use some common and understandable names instead.
format, for instance

File: packages/mobile/src/store/products/selectors.ts [hunk 20]
```diff
@@ -0,0 +1,11 @@
+import { RootState } from '~/common/types/types';
+import { createSelector } from '@reduxjs/toolkit';
+
+const selectProductById = createSelector(
+  (state: RootState) => state.products.products,
+  (state: RootState, productId: string) => productId,
+
+  (products, productId) => products.find((item) => item.id === productId),
+);
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/store/products/selectors.ts:9]:
```suggestion
  [
      (state: RootState) => state.products.products,
      (state: RootState, productId: string) => productId
  ],
  (products, productId) => products.find((item) => item.id === productId),
);
```

Also this part - `(state: RootState) => state.products.products,` - could be extracted to a separate selector, e.g. `selectProducts`

**Review by AleksandrTermenzhy — CHANGES_REQUESTED** [review]

File: packages/mobile/src/index.tsx [hunk 21]
```diff
@@ -8,10 +8,12 @@ import { store } from '~/store/store';
...
 
 i18();
 
 const App: FC = () => {
+  dayjsInit();
```
[the remark this answers is not in the document]
Comment (mazurGit) [reply packages/mobile/src/index.tsx:16]: fixed

File: packages/mobile/src/screens/home/components/product/product.tsx (hunk 16 above)
Comment (mazurGit) [reply packages/mobile/src/screens/home/components/product/product.tsx:43]: Extracted

File: packages/mobile/src/helpers/date/get-time-to-event/get-time-to-event.helper.ts (hunk 17 above)
Comment (mazurGit) [reply packages/mobile/src/helpers/date/get-time-to-event/get-time-to-event.helper.ts:6]: changed

File: packages/mobile/src/index.tsx (hunk 18 above)
Comment (mazurGit) [reply packages/mobile/src/index.tsx:11]: done

File: packages/mobile/src/helpers/helpers.ts (hunk 19 above)
Comment (mazurGit) [reply packages/mobile/src/helpers/helpers.ts:2]: changed

File: packages/mobile/src/store/products/selectors.ts (hunk 20 above)
Comment (mazurGit) [reply packages/mobile/src/store/products/selectors.ts]: solved

File: packages/mobile/src/screens/home/components/product/product.tsx [hunk 22]
```diff
@@ -0,0 +1,89 @@
+import React, { FC } from 'react';
+import { useTranslation, useAppSelector, useCustomTheme } from '~/hooks/hooks';
+import { selectProductById } from '~/store/products/selectors';
+import { Button, Image, Text, View } from '~/components/components';
+import { globalStyles } from '~/styles/styles';
+import { ProductType, ProductDto } from '@vse-bude/shared';
+import { formatPrice, getTimeToEvent } from '~/helpers/helpers';
+import { styles } from './styles';
+import { TimeWindow } from './components/time-window/time-window';
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/home/components/product/product.tsx:9]:
```suggestion
import { TimeWindow } from './components/components';
```

File: packages/mobile/src/screens/home/components/product/product.tsx [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+
+type Props = {
+  productId: string;
+};
+
+const Product: FC<Props> = ({ productId }) => {
+  const { imageLinks, title, description, price, type, endDate } =
+    useAppSelector((state) =>
+      selectProductById(state, productId),
+    ) as ProductDto;
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/home/components/product/product.tsx:19]: Why do we need explicit assertion here?

File: packages/mobile/src/screens/home/components/product/product.tsx [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+  const duration = getTimeToEvent(endDate);
+  const auction = ProductType.AUCTION;
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/home/components/product/product.tsx:21]:
```suggestion
  const isAuction = type === ProductType.AUCTION;
```

File: packages/mobile/src/screens/home/components/product/product.tsx [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+  const { colors } = useCustomTheme();
+  const { t } = useTranslation();
+
+  return (
+    <View style={[styles.container, globalStyles.boxShadow]}>
+      <View style={[styles.imgWrapper, { backgroundColor: colors.card }]}>
+        <Image source={{ uri: imageLinks[0] }} style={styles.img} />
+
+        {type === auction && (
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/screens/home/components/product/product.tsx:30]:
```suggestion
        {isAuction && (
```

File: packages/mobile/src/store/products/selectors.ts [hunk 26]
```diff
@@ -0,0 +1,14 @@
+import { RootState } from '~/common/types/types';
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/store/products/selectors.ts:1]: This should be the last one

File: packages/mobile/src/helpers/dto-name-modifiers/format-price/format-price.ts [hunk 27]
```diff
@@ -0,0 +1,7 @@
+const formatPrice = (price: number | string): string => {
+  price = price.toString();
```
[the remark this answers is not in the document]
Comment (mazurGit) [reply packages/mobile/src/helpers/dto-name-modifiers/format-price/format-price.ts:2]: fixed

File: packages/mobile/src/screens/home/components/product/product.tsx (hunk 22 above)
Comment (mazurGit) [reply packages/mobile/src/screens/home/components/product/product.tsx:9]: changed

File: packages/mobile/src/screens/home/components/product/product.tsx (hunk 23 above)
Comment (mazurGit) [reply packages/mobile/src/screens/home/components/product/product.tsx:19]: changed

File: packages/mobile/src/screens/home/components/product/product.tsx (hunk 24 above)
Comment (mazurGit) [reply packages/mobile/src/screens/home/components/product/product.tsx:21]: changed

File: packages/mobile/src/screens/home/components/product/product.tsx (hunk 25 above)
Comment (mazurGit) [reply packages/mobile/src/screens/home/components/product/product.tsx:30]: changed

File: packages/mobile/src/store/products/selectors.ts (hunk 26 above)
Comment (mazurGit) [reply packages/mobile/src/store/products/selectors.ts:1]: changed

**Review by AleksandrTermenzhy — APPROVED** [review]

**Pull request merged**